components/sidebar/sidebar_service_unittest.cc

```diff
@@ -729,4 +877,175 @@ TEST_F(SidebarServiceTestWithPlaylist, GetDefaultPanelItem) {
+           dbt_index < default_btin_types_count) {
+      if (only_builtin_types[srv_items_index].built_in_item_type ==
+          defined_order[dbt_index]) {
+        srv_items_index++;
+      }
+      dbt_index++;
+    }
+    return srv_items_index == srv_items_count;
+  }
+
+  void LoadFromPrefsTest(
+      base::Value::Dict sidebar_prefs,
+      const std::vector<SidebarItem::BuiltInItemType>& defined_order,
+      const size_t& expected_items_loaded) {
+    // base::Value::Dict
+    // sidebar(ParseTestJson(sidebar_all_builtin_visible_json));
+    GetPrefs()->Set(kSidebarItems,
+                    std::move(sidebar_prefs.Find("sidebar_items")->Clone()));
+    GetPrefs()->SetInteger(kSidebarShowOption, 0);
+    EXPECT_EQ(0, GetPrefs()->GetInteger(kSidebarShowOption));
+
+    InitSideBarService();
+
+    EXPECT_EQ(expected_items_loaded, GetSidebarService()->items().size());
+    EXPECT_TRUE(ValidateBuiltInTypesOrdering(defined_order))
+        << "Wrong order detected";
+  }
+
+ private:
+#if BUILDFLAG(ENABLE_AI_CHAT)
+  base::test::ScopedFeatureList scoped_feature_list_;
+#endif  // BUILDFLAG(ENABLE_AI_CHAT)
+  content::BrowserTaskEnvironment task_environment_{
+      base::test::TaskEnvironment::TimeSource::MOCK_TIME};
+  std::unique_ptr<TestingProfile> profile_;
+  raw_ptr<SidebarService> sidebar_service_;
+  NiceMock<MockSidebarServiceObserver> observer_;
+};
+
+TEST_F(SidebarServiceOrderingTest, BuiltInItemsDefaultOrder) {
+  InitSideBarService();
+  EXPECT_EQ(
+#if BUILDFLAG(ENABLE_AI_CHAT)
+      5UL,
+#else
+      4UL,
+#endif  // BUILDFLAG(ENABLE_AI_CHAT)
+      GetSidebarService()->items().size());
+  EXPECT_EQ(0UL, GetSidebarService()->GetHiddenDefaultSidebarItems().size());
+
+  EXPECT_TRUE(
+      ValidateBuiltInTypesOrdering({SidebarItem::BuiltInItemType::kBraveTalk,
+                                    SidebarItem::BuiltInItemType::kWallet,
+                                    SidebarItem::BuiltInItemType::kChatUI,
+                                    SidebarItem::BuiltInItemType::kBookmarks,
+                                    SidebarItem::BuiltInItemType::kReadingList,
+                                    SidebarItem::BuiltInItemType::kHistory,
+                                    SidebarItem::BuiltInItemType::kPlaylist}));
+}
+
+TEST_F(SidebarServiceOrderingTest, LoadFromPrefsAllBuiltInVisible) {
+  base::Value::Dict sidebar(ParseTestJson(sidebar_all_builtin_visible_json));
+  LoadFromPrefsTest(std::move(sidebar),
+                    {
+                        SidebarItem::BuiltInItemType::kChatUI,
+                        SidebarItem::BuiltInItemType::kWallet,
+                        SidebarItem::BuiltInItemType::kReadingList,
+                        SidebarItem::BuiltInItemType::kBookmarks,
+                        SidebarItem::BuiltInItemType::kBraveTalk,
+                    },
+#if BUILDFLAG(ENABLE_AI_CHAT)
+                    8UL
+#else
+                    7UL
+#endif  // BUILDFLAG(ENABLE_AI_CHAT)
```

Comment:
There's a lot of manual specifying here, and a lot of hard-coded counts and lists that will need changing when we add or modify sidebar item defaults. But perhaps that's what we want for tests? Would be interested to see @simonhong's opinion.

Reply:
Hmm, I think this is fine as it's based on preset data such as `sidebar_all_builtin_visible_json` and it could be used for checking preset is loaded properly.
If default item list change affects this preset data, both (preset data and this count number) should be updated.
I would suggest that it would be more better to have another constants like `sidebar_all_builtin_visible_json_total_item_count` instead of using `8UL` or `7UL`.

```suggestion
#if BUILDFLAG(ENABLE_AI_CHAT)
                    sidebar_all_builtin_visible_json_total_item_count
#else
                    sidebar_all_builtin_visible_json_total_item_count - 1
#endif  // BUILDFLAG(ENABLE_AI_CHAT)
```

Reply:
done